include/coap2/coap_pdu_internal.h

```diff
@@ -0,0 +1,224 @@
+*
+* @param proto  Session's protocol
+* @param data   The first byte of raw data to parse as CoAP PDU.
+*
+* @return       A value greater than zero on success or @c 0 on error.
+*/
+size_t coap_pdu_parse_header_size(coap_proto_t proto,
+                                 const uint8_t *data);
+
+/**
+ * Parses @p data to extract the message size.
+ * @p length must be at least coap_pdu_parse_header_size(proto, data).
+ * This function returns @c 0 on error or a number greater than zero on success.
+ *
+ * @param proto  Session's protocol
+ * @param data   The raw data to parse as CoAP PDU.
+ * @param length The actual size of @p data.
+ *
+ * @return       A value greater than zero on success or @c 0 on error.
+ */
+size_t coap_pdu_parse_size(coap_proto_t proto,
+                           const uint8_t *data,
+                           size_t length);
+
+/**
+ * Decode the protocol specific header for the specified PDU.
+ * @param pdu A newly received PDU.
+ * @param proto The target wire protocol.
+ * @return 1 for success or 0 on error.
+ */
+
+int coap_pdu_parse_header(coap_pdu_t *pdu, coap_proto_t proto);
+
+/**
+ * Verify consistency in the given CoAP PDU structure and locate the data.
+ * This function returns @c 0 on error or a number greater than zero on
+ * success.
+ * This function only parses the token and options, up to the payload start
+ * marker.
+ *
+ * @param pdu     The PDU structure to.
+ *
+ * @return       1 on success or @c 0 on error.
+ */
+int coap_pdu_parse_opt(coap_pdu_t *pdu);
+
+/**
+* Parses @p data into the CoAP PDU structure given in @p result.
+* The target pdu must be large enough to
+* This function returns @c 0 on error or a number greater than zero on success.
+*
+* @param proto   Session's protocol
+* @param data    The raw data to parse as CoAP PDU.
+* @param length  The actual size of @p data.
+* @param pdu     The PDU structure to fill. Note that the structure must
+*                provide space to hold at least the token and options
+*                part of the message.
+*
+* @return       1 on success or @c 0 on error.
+*/
+int coap_pdu_parse(coap_proto_t proto,
+                   const uint8_t *data,
+                   size_t length,
+                   coap_pdu_t *pdu);
+
+/**
+ * Clears any contents from @p pdu and resets @c used_size,
+ * and @c data pointers. @c max_size is set to @p size, any
+ * other field is set to @c 0. Note that @p pdu must be a valid
+ * pointer to a coap_pdu_t object created e.g. by coap_pdu_init().
+ */
+void coap_pdu_clear(coap_pdu_t *pdu, size_t size);
+
+/**
+ * Removes option of given type from the @p pdu.
+ */
+int coap_remove_option(coap_pdu_t *pdu, coap_pdu_option_t type);
+
+/**
+ * Inserts option of given type in the @p pdu with the appropriate data.
+ * The option will be inserted in the appropriate place in the options in
+ * the pdu.
+ */
+size_t coap_insert_option(coap_pdu_t *pdu, coap_pdu_option_t type,
```

Comment:
I thought about this for some time.  They do have an overhead of consequence moving stuff around but are needed for the RFC7959 block handling and so was inclined to discourage it for general use (can't think of a good reason why an application would want to do this).  If adding an option, it will invoke coap_insert_option() if needed.  When generally setting up a PDU, use should be made of coap_new_optlist() etc. which are sorted before adding to the PDU.

The same question is true for coap_update_token() and coap_remove_token().

Happy to go either way though here.

Reply:
True. It sounds like a good idea to make it private for now; if we really see a need in the future to make it public it will be less intrusive than the other way round.